shared/src/sources/balancer/graph_api.rs

```diff
@@ -179,23 +247,59 @@ mod pools_query {
 
             let token_count = self.tokens.len();
             self.tokens.iter().try_fold(
-                RegisteredWeightedPool {
+                RegisteredPool::Weighted(RegisteredWeightedPool {
                     pool_id: self.id,
                     pool_address: self.address,
                     tokens: Vec::with_capacity(token_count),
                     normalized_weights: Vec::with_capacity(token_count),
                     scaling_exponents: Vec::with_capacity(token_count),
                     block_created: block_created_upper_bound,
+                }),
+                |mut registered_pool, token| {
+                    if let RegisteredPool::Weighted(ref mut pool) = registered_pool {
+                        pool.tokens.push(token.address);
+                        pool.normalized_weights.push(token.weight);
+                        pool.scaling_exponents
+                            .push(18u8.checked_sub(token.decimals).ok_or_else(|| {
+                                anyhow!("unsupported token with more than 18 decimals")
+                            })?);
+                        Ok(registered_pool)
+                    } else {
+                        unreachable!("Not a weighted pool");
+                    }
```

Comment:
Instead of creating the RegisteredPool enum immediately and then needing this unreachable branch we can do:
```rust
self.tokens
    .iter()
    .try_fold(
        RegisteredWeightedPool {
            pool_id: self.id,
            pool_address: self.address,
            tokens: Vec::with_capacity(token_count),
            normalized_weights: Vec::with_capacity(token_count),
            scaling_exponents: Vec::with_capacity(token_count),
            block_created: block_created_upper_bound,
        },
        |mut pool, token| {
            pool.tokens.push(token.address);
            pool.normalized_weights.push(token.weight);
            pool.scaling_exponents
                .push(18u8.checked_sub(token.decimals).ok_or_else(|| {
                    anyhow!("unsupported token with more than 18 decimals")
                })?);
            Ok(pool)
        },
    )
    .map(RegisteredPool::Weighted)
```

And the same change for stable pools below.